pkg/controller/tas/topology_ungater.go

```diff
@@ -0,0 +1,300 @@
+}
+
+func (r *topologyUngater) Reconcile(ctx context.Context, req reconcile.Request) (reconcile.Result, error) {
+	log := ctrl.LoggerFrom(ctx).WithValues("workload", req.NamespacedName.Name)
+	log.V(2).Info("Reconcile Topology Ungater")
+
+	wl := &kueue.Workload{}
+	if err := r.client.Get(ctx, req.NamespacedName, wl); err != nil {
+		if client.IgnoreNotFound(err) != nil {
+			return reconcile.Result{}, err
+		}
+		log.Info("workload not found")
+		return reconcile.Result{}, nil
+	}
+	if wl.Status.Admission == nil {
+		log.Info("workload is not admitted")
+		return reconcile.Result{}, nil
+	}
+
+	allToUngate := make([]podWithUngateInfo, 0)
+	for _, psa := range wl.Status.Admission.PodSetAssignments {
+		if psa.TopologyAssignment != nil {
+			toUngate, err := r.podsetPodsToUngate(ctx, log, wl, &psa)
+			if err != nil {
+				log.Error(err, "failed to identify pods to ungate", "podset", psa.Name, "count", psa.Count)
+				return reconcile.Result{}, err
+			} else {
+				log.Info("identified pods to ungate for podset", "podset", psa.Name, "count", len(toUngate))
+				allToUngate = append(allToUngate, toUngate...)
+			}
+		}
+	}
+	var err error
+	if len(allToUngate) > 0 {
+		log.V(2).Info("identified pods to ungate", "count", len(allToUngate))
+		err = parallelize.Until(ctx, len(allToUngate), func(i int) error {
+			podWithUngateInfo := &allToUngate[i]
+			e := utilclient.Patch(ctx, r.client, podWithUngateInfo.pod, true, func() (bool, error) {
+				log.V(3).Info("ungating pod", "pod", klog.KObj(podWithUngateInfo.pod), "nodeLabels", podWithUngateInfo.nodeLabels)
+				utilpod.Ungate(podWithUngateInfo.pod, kueuealpha.TopologySchedulingGate)
+				if podWithUngateInfo.pod.Spec.NodeSelector == nil {
+					podWithUngateInfo.pod.Spec.NodeSelector = make(map[string]string)
+				}
+				for labelKey, labelValue := range podWithUngateInfo.nodeLabels {
+					podWithUngateInfo.pod.Spec.NodeSelector[labelKey] = labelValue
+				}
+				return true, nil
+			})
+			if e != nil {
+				log.Error(e, "failed ungating pod", "pod", klog.KObj(podWithUngateInfo.pod))
+			}
+			return e
+		})
+		if err != nil {
+			return reconcile.Result{}, err
+		}
+	}
+	return reconcile.Result{}, nil
+}
+
+func (r *topologyUngater) Create(event event.CreateEvent) bool {
+	wl, isWl := event.Object.(*kueue.Workload)
+	if isWl {
+		return isTASWorkload(wl)
+	}
+	return true
+}
+
+func (r *topologyUngater) Delete(event event.DeleteEvent) bool {
+	wl, isWl := event.Object.(*kueue.Workload)
+	if isWl {
+		return isTASWorkload(wl)
+	}
+	return true
+}
+
+func (r *topologyUngater) Update(event event.UpdateEvent) bool {
+	_, isOldWl := event.ObjectOld.(*kueue.Workload)
+	newWl, isNewWl := event.ObjectNew.(*kueue.Workload)
+	if isOldWl && isNewWl {
+		return isTASWorkload(newWl)
+	}
+	return true
+}
+
+func isTASWorkload(wl *kueue.Workload) bool {
+	if wl.Status.Admission == nil {
```

Comment:
```suggestion
	if wl.Status.Admission == nil || slices.ContainsFunc(wl.Status.Admission.PodSetAssignments, func(psa kueue.PodSetAssignment) bool { return psa.TopologyAssignment != nil } ) {
```

Reply:
cleaned up along the lines in a helper function `isAdmittedByTAS`